internal/daemon/vendor-specific-plugins/marvell/main.go

```diff
@@ -33,19 +34,172 @@ const (
 	Version      string = "0.0.1"
 )
 
+var (
+	mrvlVspServerInstance *mrvlVspServer
+)
+
+type deviceConfig struct {
+	PortType      string `yaml:"portType"`
+	NoOfPortPairs int    `yaml:"noOfPortPairs"`
+}
+type mrvlDeviceInfo struct {
+	nfInterfaceName string
+	dpInterfaceName string
+	dpMAC           string
+	health          string
+}
 type mrvlVspServer struct {
 	pb.UnimplementedLifeCycleServiceServer
 	pb.UnimplementedNetworkFunctionServiceServer
 	pb.UnimplementedDeviceServiceServer
 	opi.UnimplementedBridgePortServiceServer
-	log         logr.Logger
-	grpcServer  *grpc.Server
-	wg          sync.WaitGroup
-	done        chan error
-	startedWg   sync.WaitGroup
-	pathManager utils.PathManager
-	version     string
-	isDPUMode   bool
+	log           logr.Logger
+	grpcServer    *grpc.Server
+	wg            sync.WaitGroup
+	done          chan error
+	startedWg     sync.WaitGroup
+	pathManager   utils.PathManager
+	version       string
+	isDPUMode     bool
+	deviceStore   map[string]mrvlDeviceInfo
+	portType      string
+	noOfPortPairs int
+}
+
+// readConfig function to read the config file for the given device
+func readConfig() (deviceConfig, error) {
+	config_path := "/config.yaml"
+	data, err := os.ReadFile(config_path)
+	if err != nil {
+		return deviceConfig{}, err
+	}
+	var config deviceConfig
+	err = yaml.Unmarshal(data, &config)
+	if err != nil {
+		return deviceConfig{}, err
+	}
+	return config, nil
+}
+
+// createVethPair function to create a veth pair with the given index and InterfaceInfo
+func (vsp *mrvlVspServer) createVethPair(index int) error {
+	nfInterfaceName := fmt.Sprintf("nf_interface%d", index)
+	dpInterfaceName := fmt.Sprintf("dp_interface%d", index)
+	vethLink := &netlink.Veth{
+		LinkAttrs: netlink.LinkAttrs{Name: nfInterfaceName},
+		PeerName:  dpInterfaceName,
+	}
+	if err := netlink.LinkAdd(vethLink); err != nil {
+		return err
+	}
+	if err := netlink.LinkSetUp(vethLink); err != nil {
```

Comment:
Below we call `LinkSetUp()` on the `peerLink` instance that we just looked up fresh with `LinkByName()`.

I think that should also be done in this case. I mean, move this before `netlink.LinkSetUp(peerLink)` and operate on `nfLink`.

Otherwise, `LinkSetUp(vethLink)` needs internally call `LinkByName()` to find the ifindex. Instead, by using `LinkSetUp(nfLink)`, the ifindex is already known.

Reply:
Okay so if i do LinkSetUp() on nflink after LinkByName() would be sufficient No need for vethLink 
is that what you mean?

Reply:
`vethLink` is still needed for `LinkAdd()`.

I mean
```diff
From ebe1cff131ac5e451f8529cd61b18de10fc48c42 Mon Sep 17 00:00:00 2001
From: Thomas Haller <thaller@redhat.com>
Date: Tue, 1 Oct 2024 11:26:48 +0200
Subject: [PATCH 1/1] fixup! Add GetDevice() to Marvell VSP This commit adds
 GetDevice Support for mrvl-vsp

---
 internal/daemon/vendor-specific-plugins/marvell/main.go | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/internal/daemon/vendor-specific-plugins/marvell/main.go b/internal/daemon/vendor-specific-plugins/marvell/main.go
index 51ba3c7ab12b..d7c1c7ab9659 100644
--- a/internal/daemon/vendor-specific-plugins/marvell/main.go
+++ b/internal/daemon/vendor-specific-plugins/marvell/main.go
@@ -85,34 +85,35 @@ func readConfig() (deviceConfig, error) {
 func (vsp *mrvlVspServer) createVethPair(index int) error {
 	nfInterfaceName := fmt.Sprintf("nf_interface%d", index)
 	dpInterfaceName := fmt.Sprintf("dp_interface%d", index)
 	vethLink := &netlink.Veth{
 		LinkAttrs: netlink.LinkAttrs{Name: nfInterfaceName},
 		PeerName:  dpInterfaceName,
 	}
 	if err := netlink.LinkAdd(vethLink); err != nil {
 		return err
 	}
-	if err := netlink.LinkSetUp(vethLink); err != nil {
-		return err
-	}
 
 	nfLink, err := netlink.LinkByName(nfInterfaceName)
 	if err != nil {
 		return err
 	}
 
 	peerLink, err := netlink.LinkByName(dpInterfaceName)
 	if err != nil {
 		return err
 	}
 
+	if err := netlink.LinkSetUp(nfLink); err != nil {
+		return err
+	}
+
 	if err := netlink.LinkSetUp(peerLink); err != nil {
 		return err
 	}
 
 	vsp.deviceStore[nfLink.Attrs().HardwareAddr.String()] = mrvlDeviceInfo{
 		nfInterfaceName: nfInterfaceName,
 		dpInterfaceName: dpInterfaceName,
 		dpMAC:           peerLink.Attrs().HardwareAddr.String(),
 		health:          "Healthy",
 	}
-- 
2.46.2

```